include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp

```diff
@@ -231,112 +239,183 @@ class alignment_executor_two_way
-            current_chunk_size = std::ranges::distance(current_chunk);
-            assert(in_avail() >= current_chunk_size);
-
-            exec_handler.execute(kernel, std::move(current_chunk), [write_to = gptr] (auto && alignment_result) mutable
+            exec_handler.execute(kernel, *resource_it, [target_bucket_iter] (auto && alignment_result)
             {
-                *write_to = std::move(alignment_result);
-                ++write_to;
+                target_bucket_iter->push_back(std::move(alignment_result));
             });
         }
 
         exec_handler.wait();
 
-        // Update the available get positions to cover the part of the buffer that was filled.
-        setg(std::ranges::begin(buffer), std::ranges::begin(buffer) + count);
+        // Move to the first non empty bucket.
+        find_next_filled_bucket();
+
+        if (is_buffer_empty())
+            return underflow_status::empty_buffer;
 
-        return in_avail();
+        return underflow_status::filled_buffer;
     }
     //!\}
 
     /*!\name Miscellaneous
      * \{
      */
 
-    /*!\brief Initialises the underlying buffer.
-     * \param[in] size The initial size of the buffer.
-     * \param[in] init_value The value used to initialise the buffer.
+    /*!\brief Whether the internal buffer is empty.
+     * \returns `true` if all elements of the internal buffer have been consumed, otherwise `false`.
+     */
+    bool is_buffer_empty() const
+    {
+        return bucket_iterator == bucket_end || gptr == bucket_iterator->end();
+    }
+
+    /*!\brief Resets the buckets.
+     *
+     * \details
+     *
+     * Clears all buckets and sets the bucket iterator to the first iterator, such that the allocated memory for each
+     * bucket can be reused between invocations of seqan3::detail::alignment_executor_two_way::underflow.
      */
-    void init_buffer(size_t const size, value_t const init_value)
+    void reset_buckets()
     {
-        buffer.resize(size, std::move(init_value));
-        setg(std::ranges::end(buffer), std::ranges::end(buffer));
+        // Clear all buckets
+        for (auto & bucket : bucket_container)
+            bucket.clear();
+
+        // Reset the iterators over the buckets.
+        bucket_iterator = bucket_container.begin();
+        bucket_end = bucket_iterator + available_buckets();
+    }
+
+    /*!\brief Returns the number of remaining buckets.
+     *
+     * \details
+     *
+     * The number of available buckets is determined by the minimum of the total number of available buckets and the
+     * remaining number of elements of the underlying resource.
+     */
+    std::iter_difference_t<resource_iterator_type> available_buckets()
+    {
+        using difference_t = std::iter_difference_t<resource_iterator_type>;
+        return std::min<difference_t>(bucket_container_size,
+                                      std::ranges::distance(resource_it, std::ranges::end(resource)));
```

Comment:
> What happens if my bucket count is smaller than the resources that needs to be processed?

Only bucket count many elements are filled within one call of underflow. 

> This is weird? There should always be enough buckets for the amount of work I give.
I think this would be an illegal state.
I think an invariant of this data structure should be size(buckets) > size (resources) and therefore bucket_end = bucket_begin + size(resources). There might be more capacity in the vector, but you only use resources many size. Why don't we use the capacity facility of the vector and resize it to the actual size which might shrink or grow the vector? And we can always use the end iterator of the buckets?

No, why? I am not sure I understand your point here. I mean the change to the bucket interface is necessary to allow a single invocation of the algorithm to produce zero, one or more than one result. This wasn't possible before. Now, the buckets are the buffer size like before. So at some point (of course irrespective of the other proposal of yours) if the buffer should be 6 for example you can only process 6 many elements from the resource in one call to underflow. But the last chunk of the resource might have less than 6 elements. so in that case you want to make the buffer smaller. I can do this in many different ways but your solution would also imply checking the remaining size of the resource. That could however hurt in the runtime if distance is linear. So instead I change it that the loop inside of underflow keeps track of the number of filled buckets in what ever comes first. THe end of the resource or the end of the bucket count.